### PR TITLE
[Intel MKL] Optimized implementation of GatherND using OpenMP

### PR DESCRIPTION
**Pull request by nhasabni**
Eigen version does not parallelize well. With my parallel version, I get almost 2X improvement.

### CODE REVIEW

**Comment by qlzh727** [comment]
Nudge for review.

File: tensorflow/core/kernels/gather_nd_op_cpu_impl.h [hunk 1]
```diff
@@ -113,10 +113,25 @@ struct GatherNdSlice<CPUDevice, T, Index, IXDIM> {
...
+    // Additionally, since it uses scalar (Tscratch) to invoke 'generate', it
+    // needs to go through redundant operations like 'reshape', 'broadcast' and
+    // 'sum'. OpenMP loop below essentially does same thing as Eigen code, but
+    // is considerably more efficient.
+    #pragma omp parallel for
+    for (Eigen::DenseIndex i = 0; i < batch_size; i++) {
+      const Eigen::array<Eigen::DenseIndex, 1> loc = i;
+      gather_nd_generator(loc);
+    }
+#else
```
Comment (ebrevdo) [inline tensorflow/core/kernels/gather_nd_op_cpu_impl.h]: thanks!  if you have time, can you also make a non-intel mkl modification below to use Shard(), like [this example](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/reverse_op.cc#L81)?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @qlzh727: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/core/kernels/gather_nd_op_cpu_impl.h (hunk 1 above)
Comment (nhasabni) [reply tensorflow/core/kernels/gather_nd_op_cpu_impl.h]: @ebrevdo Sorry, but I don't think I have time for that change. Context switch to make this small change is going to be costly for me. Sorry again.

**Comment by qlzh727** [comment]
@nhasabni, is this PR still alive? The sanity check is failing for this PR. Please take a look.

**Comment by nhasabni** [comment]
@qlzh727 Yes, this PR is still active. I will look into the failure.

**Comment by ebrevdo** [comment]
It's ok to skip adding the shard non-openmp version.

**Comment by nhasabni** [comment]
@qlzh727 Is it possible to restart the tests? I looked into the CI failure, I do not think it is related to my change. It was in `do_bazel_nobuild: bazel nobuild`.

**Comment by nhasabni** [comment]
@qlzh727 Thanks for kicking off the run. I see 2 failures reported, but I am not able to access any of the logs files for the failures.

**Comment by qlzh727** [comment]
The clang formatting seems to be failing, can u update your code and fix that?

**Comment by qlzh727** [comment]
Can u access that link?

https://source.cloud.google.com/results/invocations/3707dc61-9a5a-4b9c-9f9a-277c1f4c6145/targets/%2F%2Ftensorflow%2Ftools%2Fci_build:gen_ci_clang_format_out/log

**Comment by nhasabni** [comment]
@qlzh727 Yes I can. But I don't see what are cpplint errors. All that I see is my change and the command that was executed. I don't see any output log.

**Comment by qlzh727** [comment]
There is a diff of the original file and the output of clang, you need to update your file to match the clang output, otherwise it will complain for the formatting.

**Comment by nhasabni** [comment]
@qlzh727 So all tests look good now?

**Comment by qlzh727** [comment]
@ebrevdo, please approve the PR if it LG to you. I will then sync and merge it.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @qlzh727: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by qlzh727** [comment]
@ebrevdo is on holiday at the moment, will ping him to approve the change once he's back.

**Review by ebrevdo — APPROVED** [review]

**Pull request merged**